Fixes #24652 - subnet_for returns highest CIDR prefix

lzap opened the pull request:
It returns randomly sorted by VLANID which can be used as a workaround.
This patch sorts them by CIDR prefix, which we unfortunately don'ŧ store
in the database so we need to sort in-memory.

Proper solution would be to store both probably, with a database index.
Let's do this in another path tho.

lzap committed:
Fixes #24652 - subnet_for returns highest CIDR prefix

ekohl commented on app/models/subnet.rb:
```diff
@@ -273,13 +273,13 @@ def supported_ipam_modes_with_translations
       return unless ip.present?
       ip = IPAddr.new(ip)
-      Subnet.all.detect {|s| s.family == ip.family && s.contains?(ip)}
+      Subnet.unscoped.all.select {|s| s.family == ip.family && s.contains?(ip)}.sort {|a,b| b.cidr <=> a.cidr}.first
```
Isn't https://ruby-doc.org/core-2.5.1/Enumerable.html#method-i-max_by better than (reverse) sorting and taking the first?

lzap replied:
This is indeed good idea, thanks! Rebased.

lzap commented:
Rebased for tests to pass due to FactoryGirl5 issue.

tbrisker commented on app/models/subnet.rb:
```diff
@@ -273,13 +273,13 @@ def supported_ipam_modes_with_translations
       return unless ip.present?
       ip = IPAddr.new(ip)
-      Subnet.all.detect {|s| s.family == ip.family && s.contains?(ip)}
+      Subnet.unscoped.all.select {|s| s.family == ip.family && s.contains?(ip)}.max_by {|x| x.cidr}
```
you can also use `.max_by(&:cidr)`, but it's not crucial.

tbrisker reviewed (changes requested):
Can you please also add a test for ipv6? 
I'm not too concerned about finding max in memory, in most cases there will only be 1-2 matching subnets. I think that's better than storing cidr in db and having to make sure it is in sync with the mask all the time.

lzap commented:
I agree on the other hand I am pretty sure there is somebody with Foreman having 10,000 subnets. Anyway, rebased the shorter syntax and added IPv6 test, good idea.

ekohl reviewed (approved):
If needed you could fall back to DB native checks like https://github.com/theforeman/foreman/pull/5623 but the existing code also operates in memory so this isn't regressing.

tbrisker commented:
@lzap there may be users with that many subnets but the max is only done for the matching subnets, the select would be done anyways in memory since I don't think we have a good way of checking if an ip is in a subnet at the db level.

lzap commented:
Katello::Api::Rhsm::CandlepinDynflowProxyControllerIntegrationTest.test_0001_params are not parsed in the controller failure not related it appears.

lzap commented:
What is the status @tbrisker or @ekohl please?

tbrisker reviewed (approved):
Thanks @lzap !
